cocos/ui/UIRichText.cpp

```diff
@@ -1492,115 +1493,158 @@ void RichText::formatText()
-    size_t stringLength = StringUtils::getCharacterCountInUTF8String(text);
+        float overstepPercent = (-leftSpaceWidth) / textRendererWidth;
+        int stringLength = static_cast<int>(text.length());
 
-    // rough estimate
-    int leftLength = stringLength * (1.0f - overstepPercent);
+        // rough estimate
+        int leftLength = stringLength * (1.0f - overstepPercent);
 
-    // The adjustment of the new line position
-    auto originalLeftSpaceWidth = _leftSpaceWidth + textRendererWidth;
-    auto leftStr = Helper::getSubStringOfUTF8String(curText, 0, leftLength);
-    label->setString(leftStr);
-    auto leftWidth = label->getContentSize().width;
-    if (originalLeftSpaceWidth < leftWidth) {
-        // Have protruding
-        for (;;) {
-            leftLength--;
-            leftStr = Helper::getSubStringOfUTF8String(curText, 0, leftLength);
-            label->setString(leftStr);
-            leftWidth = label->getContentSize().width;
-            if (leftWidth <= originalLeftSpaceWidth) {
-                break;
-            }
-            else if (leftLength <= 0) {
-                break;
+        // The adjustment of the new line position
+        std::string leftStr = text.getAsCharSequence(0, leftLength);
+        label->setString(leftStr);
+        textRendererWidth = label->getContentSize().width;
+        if (originalLeftSpaceWidth < textRendererWidth)  // Have protruding
+        {
+            while (leftLength-- > 0)
+            {
+                // try to erase a char
+                auto& ch = text.getString().at(leftLength);
+                leftStr.erase(leftStr.end() - ch._char.length(), leftStr.end());
+                label->setString(leftStr);
+                textRendererWidth = label->getContentSize().width;
+                if (textRendererWidth <= originalLeftSpaceWidth)  // is fitted
+                    break;
             }
         }
-    }
-    else if (leftWidth < originalLeftSpaceWidth) {
-        // A wide margin
-        for (;;) {
-            leftLength++;
-            leftStr = Helper::getSubStringOfUTF8String(curText, 0, leftLength);
-            label->setString(leftStr);
-            leftWidth = label->getContentSize().width;
-            if (originalLeftSpaceWidth < leftWidth) {
-                leftLength--;
-                break;
-            }
-            else if (static_cast<int>(stringLength) <= leftLength) {
-                break;
+        else if (textRendererWidth < originalLeftSpaceWidth)  // A wide margin
+        {
+            while (leftLength < stringLength)
+            {
+                // try to append a char
+                auto& ch = text.getString().at(leftLength);
+                ++leftLength;
+                leftStr.append(ch._char);
+                label->setString(leftStr);
+                textRendererWidth = label->getContentSize().width;
+                if (originalLeftSpaceWidth < textRendererWidth)  // protruded, undo add
```

Comment:
if originalLeftSpaceWidth == textRendererWidth, then it should return leftLength

Reply:
OK